pennylane/circuit_graph.py

```diff
@@ -12,305 +12,278 @@
         """
-        return [self.graph.nodes[i] for i in ops]
+        G = nx.DiGraph(self._graph.subgraph(ops))
+        return nx.dag.topological_sort(G)
 
-    def get_ops(self, ops):
-        """Given a set of operation indices, return the operation objects.
+    def ancestors_in_order(self, ops):
+        """Ancestors in a topological order.
 
         Args:
-            ops (Iterable[int]): a given set of operations labelled by integer
-                position in the queue
+            ops (Iterable[Operation]): set of operations in the circuit
 
         Returns:
-            List[Operation, Observable]: operations or observables
-            corresponding to given integer positions in the queue
+            list[Operation]: ancestors of the given operations, topologically ordered
         """
-        return [node["op"] for node in self.get_nodes(ops)]
+        # TODO returns the original temporal order, could also return an arbitrary topological order
+        #return self._in_topological_order(self.ancestors(ops))
+        return sorted(self.ancestors(ops), key=lambda x: x.queue_idx)
 
-    def get_names(self, ops):
-        """Given a set of operation indices, return the operation names.
+    def descendants_in_order(self, ops):
+        """Descendants in a topological order.
 
         Args:
-            ops (Iterable[int]): a given set of operations labelled by integer
-                position in the queue
+            ops (Iterable[Operation]): set of operations in the circuit
 
         Returns:
-            List[str]: operations or observables
-            corresponding to given integer positions in the queue
+            list[Operation]: descendants of the given operations, topologically ordered
         """
-        return [self.graph.nodes(data="name")[i] for i in ops]
+        return sorted(self.descendants(ops), key=lambda x: x.queue_idx)
+
 
     @property
     def layers(self):
-        """Identifies and returns a metadata list describing the
-        layer structure of the circuit.
+        """Identify the parametrized layer structure of the circuit.
 
-        Each layer is a named tuple containing:
+        Each :class:`Layer` is a namedtuple containing the fields
 
-        * ``op_idx`` *(list[int])*: the list of operation indices in the layer
+        * ``ops`` *(list[int])*: the list of Operations in the layer
 
-        * ``param_idx`` *(list[int])*: the list of parameter indices used within the layer
+        * ``param_inds`` *(list[int])*: the list of parameter indices used within the layer
 
         Returns:
-            list[Layer]: a list of layers
+            list[Layer]: the layers of the circuit
         """
-        # keep track of the layer number
-        layer = 0
-        layer_ops = {0: ([], [])}
+        # FIXME maybe layering should be greedier, for example [a0 b0 c1 d1] should layer as [a0 c1], [b0, d1] and not [a0], [b0 c1], [d1]
```

Comment:
That would be ideal! There are a lot of existing papers/theory on optimum layering of directed acyclic graphs. However I implemented this using a sort of 'brute force' approach. Our case only differs from the above in that we need to distinguish between parameterized nodes and 'fixed' nodes.